core/src/main/java/org/geysermc/geyser/pack/SkullResourcePackManager.java

```diff
@@ -0,0 +1,306 @@
+            zipOS.putNextEntry(entry);
+            zipOS.write(fillAttachableJson(template, skull).getBytes(StandardCharsets.UTF_8));
+            zipOS.closeEntry();
+        }
+    }
+
+    private static void addSkinTextures(ZipOutputStream zipOS) throws IOException {
+        for (Path skinPath : SKULL_SKINS.values()) {
+            ZipEntry entry = new ZipEntry("skull_resource_pack/textures/blocks/" + truncateHash(skinPath.getFileName().toString()) + ".png");
+            zipOS.putNextEntry(entry);
+            try (InputStream stream = Files.newInputStream(skinPath)) {
+                stream.transferTo(zipOS);
+            }
+            zipOS.closeEntry();
+        }
+    }
+
+    private static void fillTemplate(ZipOutputStream zipOS, String path, UnaryOperator<String> filler) throws IOException {
+        String template = FileUtils.readToString(path);
+        String result = filler.apply(template);
+        zipOS.write(result.getBytes(StandardCharsets.UTF_8));
+    }
+
+    private static String fillAttachableJson(String template, CustomSkull skull) {
+        return template.replace("${identifier}", skull.getCustomBlockData().identifier())
+                .replace("${texture}", truncateHash(skull.getSkinHash()));
+    }
+
+    private static String fillManifestJson(String template) {
+        Pair<UUID, UUID> uuids = generatePackUUIDs();
+        return template.replace("${uuid1}", uuids.first().toString())
+                .replace("${uuid2}", uuids.second().toString());
+    }
+
+    private static String fillTerrainTextureJson(String template) {
+        StringBuilder textures = new StringBuilder();
+        for (String skinHash : SKULL_SKINS.keySet()) {
+            String texture = String.format("\"geyser.%s_player_skin\":{\"textures\":\"textures/blocks/%s\"},\n", skinHash, truncateHash(skinHash));
+            textures.append(texture);
+        }
+        if (textures.length() != 0) {
+            // Remove trailing comma
+            textures.delete(textures.length() - 2, textures.length());
+        }
+        return template.replace("${texture_data}", textures);
+    }
+
+    private static Pair<UUID, UUID> generatePackUUIDs() {
+        UUID uuid1 = UUID.randomUUID();
+        UUID uuid2 = UUID.randomUUID();
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            for (int i = 0; i < 8; i++) {
+                md.update((byte) ((RESOURCE_PACK_VERSION >> (i * 8)) & 0xFF));
+            }
+            SKULL_SKINS.keySet().stream()
+                    .sorted()
+                    .map(hash -> hash.getBytes(StandardCharsets.UTF_8))
+                    .forEach(md::update);
+
+            ByteBuffer skinHashes = ByteBuffer.wrap(md.digest());
+            uuid1 = new UUID(skinHashes.getLong(), skinHashes.getLong());
+            uuid2 = new UUID(skinHashes.getLong(), skinHashes.getLong());
```

Comment:
Can we not just use 1 UUID variable?

Reply:
The generate UUIDs are different here because `getLong` doesn't return the same value every time.

https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getLong--

Reply:
Unoqwy is correct